src/charm.py

```diff
@@ -39,10 +41,124 @@ def __init__(self, *args):
+            return
+
+        # Try switchover and raise and exception if it doesn't succeed.
+        # If it doesn't happen on time, Patroni will automatically run a fail-over.
+        try:
+            self._patroni.switchover()
+            logger.info("successful switchover")
+        except SwitchoverFailedError as e:
+            logger.error(f"switchover failed with reason: {e}")
+
+    def _get_ip_by_unit(self, unit: Unit) -> str:
+        """Get the IP address of a specific unit."""
+        return self._peers.data[unit].get("private-address")
+
+    @property
+    def _hosts(self) -> set:
+        """List of the current Juju hosts.
+
+        Returns:
+            a set containing the current Juju hosts
+                with the names in the k8s pod name format
+        """
+        peers = self.model.get_relation(PEER)
+        hosts = [self.unit.name.replace("/", "-")] + [
+            unit.name.replace("/", "-") for unit in peers.units
+        ]
+        return set(hosts)
+
+    @property
+    def _patroni(self) -> Patroni:
+        """Returns an instance of the Patroni object."""
+        return Patroni(
+            self._unit_ip,
+            self._storage_path,
+            self._cluster_name,
+            self._member_name,
+            self.app.planned_units(),
+            self._peers_ips,
+            self._get_postgres_password(),
+            self._replication_password,
+        )
+
+    @property
+    def _peers_ips(self) -> List[str]:
+        """Fetch current list of peers IPs.
+
+        Returns:
+            A list of peers addresses (strings).
+        """
+        # Get all members IPs and remove the current unit IP from the list.
+        addresses = self.members_ips
+        current_unit_ip = self._unit_ip
+        if current_unit_ip in addresses:
+            addresses.remove(current_unit_ip)
+        return addresses
+
+    @property
+    def _units_ips(self) -> List[str]:
+        """Fetch current list of peers IPs.
+
+        Returns:
+            A list of peers addresses (strings).
+        """
+        # Get all members IPs and remove the current unit IP from the list.
+        addresses = [self._get_ip_by_unit(unit) for unit in self._peers.units]
+        addresses.append(self._unit_ip)
+        return addresses
+
+    @property
+    def members_ips(self) -> List[str]:
+        """Returns the list of IPs addresses of the current members of the cluster."""
+        return json.loads(self._peers.data[self.app].get("members_ips", "[]"))
+
+    def _add_to_members_ips(self, ip: str) -> None:
+        """Add one IP to the members list."""
+        self._update_members_ips(ip_to_add=ip)
+
+    def _remove_from_members_ips(self, ip: str) -> None:
+        """Remove IPs from the members list."""
+        self._update_members_ips(ip_to_remove=ip)
+
+    def _update_members_ips(self, ip_to_add: str = None, ip_to_remove: str = None) -> None:
+        """Update cluster members IPs."""
```

Comment:
nit: 
"""Update cluster member IPs on app data.

Member IPs on app data are used to determine when a unit of PostGres should be added or removed from the PostGres Cluster. 

NOTE: this function does not update the IPs on the PostGres cluster.
"""

Reply:
This description is great. Added in https://github.com/canonical/postgresql-operator/pull/10/commits/5f644d3a857874236d27ab2e0615846184f9fc8c.